### PR TITLE
[WIP] Avoid costly function evaluations for points that have been evaluated before

### PR DESCRIPTION
**Pull request by MechCoder**
Fixes https://github.com/scikit-optimize/scikit-optimize/issues/163

In the case of discrete spaces, I modified the optimization algorithm from choosing the point that has the best acquisition value to a point that has the best acquisition value and not been evaluated before.

### CODE REVIEW

**Comment by MechCoder** [comment]
ping @glouppe @betatim

File: skopt/gp_opt.py [hunk 1]
```diff
@@ -161,13 +163,15 @@ def gp_minimize(func, dimensions, base_estimator=None, acq="EI", xi=0.01,
...
     """
-    # Save call args 
+    # Save call args
     specs = {"args": copy.copy(inspect.currentframe().f_locals),
              "function": inspect.currentframe().f_code.co_name}
 
     # Check params
     rng = check_random_state(random_state)
     space = Space(dimensions)
+    is_discrete_space = all(
```
Comment (glouppe) [inline skopt/gp_opt.py]: Not sure I agree with this definition of discrete space. We should be clearer. `Integer` are discrete. Same if only one of the dimensions (not all) is `Categorical` or `Integer`.

File: skopt/gp_opt.py [hunk 2]
```diff
@@ -190,7 +194,7 @@ def gp_minimize(func, dimensions, base_estimator=None, acq="EI", xi=0.01,
     n_init_func_calls = len(x0) if y0 is None else 0
     n_total_init_calls = n_random_starts + n_init_func_calls
 
-    if n_total_init_calls <= 0:
+    if n_random_starts + len(x0) <= 0:
```
Comment (glouppe) [inline skopt/gp_opt.py]: why this change?

File: skopt/tree_opt.py [hunk 3]
```diff
@@ -19,13 +20,18 @@
 from .learning import GradientBoostingQuantileRegressor
 from .learning import RandomForestRegressor
 from .space import Space
+from .space import Categorical
+from .space import Integer
+from .utils import in2d
```
Comment (glouppe) [inline skopt/tree_opt.py]: what is this function?

File: skopt/utils.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import numpy as np
+
+def in2d(arr1, arr2):
```
Comment (glouppe) [inline skopt/utils.py]: we can find a much better name

File: skopt/gp_opt.py [hunk 5]
```diff
@@ -227,28 +231,48 @@ def gp_minimize(func, dimensions, base_estimator=None, acq="EI", xi=0.01,
...
                                           random_state=rng))
+
+            if is_discrete_space:
+                # Choose the point that gives the best acquisition
+                # value AND has not been reevaluated at before.
+                # This prevents costly function reevaluations.
+                # while exploring more of the search space.
+                X_in_Xi_transform = in2d(X, Xi_transform)
+
+                # Highly unlikely corner case.
```
Comment (glouppe) [inline skopt/gp_opt.py]: Shall we really stop? if there are remaining calls, we might still find an unvisited point.

File: skopt/gp_opt.py [hunk 6]
```diff
@@ -257,9 +281,23 @@ def gp_minimize(func, dimensions, base_estimator=None, acq="EI", xi=0.01,
...
-
-                if a < best:
-                    next_x, best = x, a
+                    X.append(x)
+                    values.append(a)
+
+            X = np.asarray(X)
+            values = np.asarray(values)
+
+            if is_discrete_space:
```
Comment (glouppe) [inline skopt/gp_opt.py]: I am not convinced this is a good strategy for `lbfgs`. It that optimization converges, it may always find the same (visited) point, even if the space has not been exhaustively searched.

**Comment by glouppe** [comment]
`dummy_minimize` should also be changed

**Comment by glouppe** [comment]
What do we do when the function is noisy? It this case, evaluating it more than once might be helpful.

File: skopt/gp_opt.py (hunk 1 above)
Comment (glouppe) [inline skopt/gp_opt.py]: Also, couldnt we rather extend the space API with a property or something?

**Comment by glouppe** [comment]
Shall we focus on that for 0.1? I would personally postpone it, as it is not as trivial as it appears.

File: skopt/gp_opt.py (hunk 6 above)
Comment (MechCoder) [inline skopt/gp_opt.py]: On a side note, not sure it is a good strategy to use `lbfgs` for these spaces. I sent a PR at https://github.com/scikit-optimize/scikit-optimize/pull/172 to fix that.

**Comment by betatim** [comment]
I agree with https://github.com/scikit-optimize/scikit-optimize/pull/170#issuecomment-236412433 so I think we should try and find some cases (and make examples for them) to study when/why this happens. If only so we can give a recommendation to the user, who will most likely be confused that it is happening and think it is a bug (like we did when we first saw it). So 👍 for post-0.1

**Comment by MechCoder** [comment]
By a noisy function, do you mean that which can have more than one value at a given point?

**Comment by MechCoder** [comment]
If we do consider those, then I would just add a warning if a point is chosen more than once and not do anything special. Let us hold the discussion off to after 0.1

**Comment by glouppe** [comment]
> By a noisy function, do you mean that which can have more than one value at a given point?

Yes, and I believe we should assume it is stochastic unless otherwise specified. (E.g. when tuning hyper-parameters of a model, the CV score is a noisy estimate of the true generalization of the model)

**Comment by betatim** [comment]
I would prefer to merge this now with code that assumes that the objective is noisy -> give a warning when a point is selected again. Then create an issue to discuss how to signal that the objective is not noisy in which case this warning might be an error. Having the warnings under the assumption of noisiness is useful and we shouldn't delay it until we figure out the API for "not a noisy objective".

(also rebase: 🔁 🎸)

**Comment by MechCoder** [comment]
Yeah cool. I'll do that tomorrow.

Btw, I think we should tune the noise hyperparameter instead of setting it to a constant here (https://github.com/scikit-optimize/scikit-optimize/blob/master/skopt/gp_opt.py#L35)

File: skopt/gp_opt.py (hunk 1 above)
Comment (betatim) [inline skopt/gp_opt.py]: How about instead of trying to decide if a space is "finite", why don't we check if a new sample is already in the list of previously evaluated samples (using `np.allclose` and friends)?
